src/logging.rs

```diff
@@ -0,0 +1,82 @@
+use log::{LevelFilter, Metadata, Record};
+use std::ffi::CString;
+use std::os::raw::c_char;
+use std::os::raw::c_int;
+
+pub type LogCallback = unsafe extern "C" fn(level: c_int, msg: *const c_char);
+
+struct LogProxy {
+    callback: LogCallback,
+}
+
+impl log::Log for LogProxy {
+    fn enabled(&self, _metadata: &Metadata) -> bool {
+        true
+    }
+
+    fn log(&self, record: &Record) {
+        if self.enabled(record.metadata()) {
+            let callback: LogCallback = self.callback;
+            let msg = record.args().to_string();
+            let msg_c = CString::new(msg).unwrap();
+            unsafe {
+                callback(record.level() as c_int, msg_c.as_ptr());
+            }
+            // We do not use std::mem::forget(s), so Rust will reclaim the memory
+            // once msg_c gets cleared. The callback should thus make a copy.
+        }
+    }
+
+    fn flush(&self) {}
+}
+
+// Store the logger instance as a mutable nullable static object
+static mut LOGGER: Option<LogProxy> = None;
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+pub enum LogLevel {
+    Off = 0,
+    Error = 1,
+    Warn = 2,
+    Info = 3,
+    Debug = 4,
+    Trace = 5,
+}
```

Comment:
@kvark is the the way to go? (I copied this from `wgpu-types`)

Reply:
Yes, this look good!